docs/testing-guidelines.md

```diff
@@ -0,0 +1,122 @@
+
+This document outlines the testing guidelines and best practices for UDS Packages. These guidelines aim to ensure consistency, reliability, and quality across all of our packages.
+
+## <a name="journey-testing"></a>Journey Testing
+
+Journey testing is crucial for ensuring that UDS packages function correctly within their deployment environment and integrate properly with other UDS components.
+
+#### <a name="definition"></a>**Definition**
+A **journey** in the context of UDS package testing is defined as:
+> The minimal and least fragile way to test all features of an application that could be broken from how we deploy it.
+
+#### <a name="key-principles"></a>**Key Principles**
+- Focus on testing the deployment, not the application itself (that's the upstream developer's job).
+- Ensure that our deployment method doesn't break key functionalities.
+- Test integration with UDS core components (e.g., Istio, Keycloak).
+
+#### <a name="implementation-guidelines"></a>**Implementation Guidelines**
+- Design tests that cover essential functionality without being overly complex.
+- Focus on deployment-related issues (e.g., network policies, SSO access, cluster resource access).
+- Handle limitations due to licensing or other constraints by documenting them and implementing what testing is possible.
+
+#### <a name="tools"></a>**Tools**
+- UI Testing: Playwright, Cypress
+- Non-UI Testing: Jest
+- Custom Scripts: Bash or other scripting languages as needed
+
+## <a name="upgrade-testing"></a>Upgrade Testing
+Ensure that the current development package works when deployed over the previously released version.
+
+#### <a name="key-considerations"></a>**Key Considerations**
+- Test data migration and persistence.
+- Verify that configurations are properly carried over or updated.
+- Check for any breaking changes in APIs or integrations.
+
+## <a name="linting-and-static-analysis"></a>Linting and Static Analysis
+
+Use appropriate tooling to ensure code quality and consistency.
+
+#### <a name="recommended-tools"></a>**Recommended Tools**
+- [YAML Linting](https://github.com/adrienverge/yamllint): `yamllint`
+- [Zarf Package Linting](https://docs.zarf.dev/commands/zarf_dev_lint/): `zarf dev lint`
+- [Shellcheck](https://www.shellcheck.net/) for any bash scripts.
+
+## <a name="best-practices"></a>**Best Practices**
+
+1. **Consistency**: Use similar testing approaches across different UDS packages where possible.
+2. **Documentation**: Clearly document what each test is checking and why.
+3. **Maintenance**: Keep tests updated as the application and deployment methods evolve.
+4. **CI/CD Integration**: Incorporate all types of tests into your CI/CD pipeline.
+5. **Error Handling**: Ensure tests fail gracefully and provide clear error messages.
+6. **Idempotency**: Ensure tests produce consistent results regardless of execution order. Design them to handle dynamic and asynchronous workloads without compromising output integrity.
```

Comment:
I think we want "repeatable" or another word that "Idempotency" which has other (pedantic) specific meaning

Reply:
Good call out! I stuck with the original section names but "Repeatability" makes more sense here.